packages/runtime/src/runtime.ts

```diff
@@ -324,19 +347,160 @@ export class Runtime {
+      sender,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      target,
+      state,
+      mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetId);
+    if (asset.freeze !== sender) {
+      throw new TealError(ERRORS.ASA.FREEZE_ERROR, { address: asset.freeze });
     }
+  }
 
-    const creatorAcc = this.assertAccountDefined(addr, this.store.accounts.get(addr));
-    const assetDef = creatorAcc.getAssetDef(assetId);
-    return this.assertAssetDefined(assetId, assetDef);
+  /**
+   * https://developer.algorand.org/docs/features/asa/#freezing-an-asset
+   * Freezes assets for a target account
+   * @param sender sender address
+   * @param assetId asset index
+   * @param freezeTarget target account
+   * @param freezeState target state
+   * @param payFlags transaction parameters
+   */
+  freezeAsset (
+    sender: string, assetId: number, freezeTarget: string,
+    freezeState: boolean, payFlags: TxParams
+  ): void {
+    this.mkAssetFreezeTx(sender, assetId, freezeTarget, freezeState, payFlags);
+    const acc = this.assertAccountDefined(freezeTarget, this.store.accounts.get(freezeTarget));
+    acc.setFreezeState(assetId, freezeState);
+  }
+
+  // creates asset revoke transaction object.
+  mkAssetRevokeTx (
+    sender: string, recipient: string, assetId: number,
+    revocationTarget: string, amount: number, flags: TxParams, assetID: number
+  ): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetTransferTxnWithSuggestedParams(
+      sender,
+      recipient,
+      flags.closeRemainderTo,
+      revocationTarget,
+      amount,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetID);
+    if (asset.clawback !== sender) {
+      throw new TealError(ERRORS.ASA.CLAWBACK_ERROR, { address: asset.clawback });
+    }
+  }
+
+  /**
+   * https://developer.algorand.org/docs/features/asa/#revoking-an-asset
+   * Revoking an asset for an account removes a specific number of the asset
+   * from the revoke target account.
+   * @param sender sender address
+   * @param recipient asset receiver address
+   * @param assetId asset index
+   * @param revocationTarget revoke target account
+   * @param amount amount of assets
+   * @param payFlags transaction parameters
+   */
+  revokeAsset (
+    sender: string, recipient: string,
+    assetID: number, revocationTarget: string,
+    amount: number, payFlags: TxParams
+  ): void {
+    this.mkAssetRevokeTx(sender, recipient, assetID, revocationTarget, amount, payFlags, assetID);
+    // Transfer assets
+    const fromAssetHolding = this.getAssetHolding(assetID, revocationTarget);
+    const toAssetHolding = this.getAssetHolding(assetID, recipient);
+    this.assertAssetNotFrozen(assetID, revocationTarget);
+    this.assertAssetNotFrozen(assetID, recipient);
```

Comment:
I don't think this is correct. I was reading the docs and found that asset `clawback` can revoke an asset **even if the asset is frozen**. Can you please confirm?

Reply:
please share the link

Reply:
couldn't find anything in this main doc https://developer.algorand.org/docs/features/asa/#revoking-an-asset

Reply:
It's mentioned here: [https://developer.algorand.org/tutorials/asa-javascript/#part-6-revoke](https://developer.algorand.org/tutorials/asa-javascript/#part-6-revoke) and here: [https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/](https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/) as well.

Reply:
ok